### PR TITLE
Add template preview button

### PR DESCRIPTION
**Pull request by sastels**
# Summary | Résumé

Add a preview screen when editing / creating templates.

https://app.zenhub.com/workspaces/notify-planning-614b3ad91bc2030015ed22f5/issues/cds-snc/notification-planning/191

https://www.figma.com/file/RLcM0JwlQsd2uVK3GJr5qY/Drafting-message-templates-%7C-Creer-des-gabarits-de-message?node-id=1365%3A326

# Test instructions | Instructions pour tester la modification
~- note that the Heroku review app cannot access Redis so this PR won't work. Should work locally (with redis running) and on staging / production.~ installed redis in Heroku 🎉 

- edit or create a template and go through the edit/create <-> preview -> save
- if a template is previewed but not saved it will not be updated / created

# Reviewers needed

A dev
- [x] code review

A11y checks 
- [x] A11y tree review on inputs
- [x] Automated checks. (ARC toolkit, Axe or wave)
- [x] Voice over with safari

### CODE REVIEW

**Comment by amazingphilippe** [comment]
I think I found a flaw in our logic:

1. Let's say I edit a template, preview it, then decide I don't want to change anything. So I go back and navigate somewhere else. 
2. The template is still saved, even if I didn't press on the save button.

Similar when creating a template:
1. I start writing my template, preview it, and decide not to save it. 
2. I'll have a surprise when I see my template in the template folder!

**Comment by sastels** [comment]
This is true. I don't think there's an easy way around this currently. Though we might be able to save some data temporarily in the session... I've done that before. Let me take a look at that...

File: app/templates/views/templates/preview_template.html [hunk 1]
```diff
@@ -0,0 +1,21 @@
+{% extends "admin_template.html" %}
+{% from "components/folder-path.html" import folder_path, page_title_folder_path %}
+{% from "components/page-footer.html" import sticky_page_footer_two_submit_buttons %}
+{% from "components/page-header.html" import page_header %}
+
+{% block maincolumn_content %}
+  <div>
+    {{ page_header(
+        template._template.name,
```
Comment (amazingphilippe) [inline app/templates/views/templates/preview_template.html:9]: Looking through the a11y audit, we have some issues about descriptive headings. I'm wondering if we could use this here and have the heading say `"Previewing template {template.name}"` or in French `"Aperçu du gabarit {template.name}"`

Comment (sastels) [reply app/templates/views/templates/preview_template.html:9]: Done. Wondering if the template name should be in quotes?
image

Comment (amazingphilippe) [reply app/templates/views/templates/preview_template.html:9]: I think this is okay. Its consistent with How I did headings in the template folders. We can revisit in bulk when we find out what is best!

**Review by jimleroyer — COMMENTED** [review]
Would there be UI tests we can add if we added a new screen? (I can't see the actual changes on the deployed instances at the moment, as I have some trouble with authenticating in staging atm).

File: app.json [hunk 2]
```diff
@@ -42,7 +42,11 @@
     }
   },
   "formation": {},
-  "addons": [],
+  "addons": [
+    {
+      "plan": "heroku-redis:hobby-dev"
```
[the remark this answers is not in the document]
Comment (sastels) [reply app.json:47]: This adds redis to heroku review apps so that the preview screen will work on them 😅

File: app/main/views/templates.py [hunk 3]
```diff
@@ -69,17 +70,60 @@ def get_email_preview_template(template, template_id, service_id):
...
+def preview_prefix(service_id, template_id=None):
+    return f"template-preview:{service_id}:{template_id}"
+
+
+def set_preview_data(data, service_id, template_id=None):
+    prefix = preview_prefix(service_id, template_id)
+    for key in preview_data_keys:
+        value = data.get(key)
+        if value is not None:
+            redis_client.set(key=f"{prefix}-{key}", value=value.encode(), ex=int(timedelta(days=1).total_seconds()))
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/main/views/templates.py:105]: Yes, I'm storing them in redis here - the flask `session` is limited to 4K in size, while we (prod services) have several templates greater than 10K in size.

File: app/main/views/templates.py [hunk 4]
```diff
@@ -69,17 +70,60 @@ def get_email_preview_template(template, template_id, service_id):
...
+    for key in preview_data_keys:
+        value = redis_client.get(f"{prefix}-{key}")
+        data[key] = value.decode() if value is not None else None
+    return data
+
+
+def delete_preview_data(service_id, template_id=None):
+    prefix = preview_prefix(service_id, template_id)
+    for key in preview_data_keys:
+        redis_client.delete(f"{prefix}-{key}")
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/main/views/templates.py:120]: That would be useful 👍

File: app/main/views/templates.py [hunk 5]
```diff
@@ -69,17 +70,60 @@ def get_email_preview_template(template, template_id, service_id):
             service_id=service_id,
             template_id=template_id,
             filetype="png",
-        ),
+        )
+        if template_id and service_id
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/main/views/templates.py:74]: yeah, I agree looking at it that it's hard to follow. Likely clearer to just compute it outside the `get_template()` function.

**Comment by amazingphilippe** [comment]
> Would there be UI tests we can add if we added a new screen? (I can't see the actual changes on the deployed instances at the moment, as I have some trouble with authenticating in staging atm).

Good question, I'm wondering if we could borrow some tests from the template page and do some checks for rendering proper html. The preview page is essentially the same as the template page, minus all the distractions.

**Review by amazingphilippe — CHANGES_REQUESTED** [review]
Just one or two things to change for a11y:
- [x] Prepend the page heading to the page title on the preview page
- [ ] Maybe this would be extra: We should have a lang attribute on the email preview, and everywhere we preview a message. If too extra, we should create another issue for this.

File: app/main/views/templates.py [hunk 6]
```diff
@@ -69,17 +70,60 @@ def get_email_preview_template(template, template_id, service_id):
...
+        else None,
         show_recipient=True,
         page_count=get_page_count_for_letter(template),
     )
 
     return email_preview_template
 
 
+preview_data_keys = [
```
[the remark this answers is not in the document]
Comment (sastels) [reply app/main/views/templates.py]: refactored the code to not need that constant

**Comment by jimleroyer** [comment]
I thought in the beginning we'd use caching on the client side for some specific reason but my assumptions probably were wrong. Why do we cache the template preview in the end? Asking to clear out my confusion around this.

**Comment by sastels** [comment]
> I thought in the beginning we'd use caching on the client side for some specific reason but my assumptions probably were wrong. Why do we cache the template preview in the end? Asking to clear out my confusion around this.

- caching it so we can go to the preview page without saving the template in the database.
- I looked at adding the template to the client side `session` cache but there's a 4K limit on how big that can be.

**Comment by sastels** [comment]
Added more tests around the new preview page. Note that rendering the template is done by ` notifications_utils` code and is tested there. Here I've added tests for the stuff around the template.

**Comment by sastels** [comment]
> * Maybe this would be extra: We should have a lang attribute on the email preview, and everywhere we preview a message. If too extra, we should create another issue for this.

As discussed IRL we'll create a new issue to investigate this (basically: what lang tag should we use?)

File: app/main/views/templates.py (hunk 4 above)
[the remark this answers is not in the document]
Comment (sastels) [reply app/main/views/templates.py]: I looked into this a little and I think they don't really apply / would make the code more confusing to try to shoe-horn them in. For example, the `@cache.set` decorator returns the cached value, and if it's not there calls the function and then sets the cache value. In this PR, in one spot we want to set the key regardless if it already exists, and in another spot we want to get the value of the key (and if it's not set it's an error).

I did steal some code from `cache` though, that simplified things substantially - instead of setting the keys individually, I now just take the data dict, json it, and set one key to the json string.

**Review by jimleroyer — APPROVED** [review]
LGTM -- if everything is fine with @amazingphilippe requests

**Review by amazingphilippe — APPROVED** [review]
👍
